lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerProcess.scala

```diff
@@ -0,0 +1,327 @@
+    * actually been started and contains a handle for that process; the second
+    * one is completed with the exit code once the process exits. The second
+    * future may also be completed with an exception if the process cannot be
+    * started, in such case the first future never completes.
+    */
+  private def launchProcessInBackground(): (Future[Process], Future[Int]) = {
+    val processStartPromise = Promise[Process]()
+    val stopped = Future {
+      runServer(processStartPromise)
+    }(LanguageServerProcess.forkedProcessExecutionContext)
+
+    (processStartPromise.future, stopped)
+  }
+
+  /** Waits for the process to actually start (this may take a long time if
+    * there are locked locks).
+    *
+    * Once the process is started, it goes to the bootingStage.
+    */
+  private def startingStage: Receive = {
+    case ProcessStarted(process) =>
+      import context.dispatcher
+      val cancellable =
+        context.system.scheduler.scheduleOnce(bootTimeout, self, TimedOut)
+      context.become(bootingStage(process, cancellable))
+      unstashAll()
+      self ! AskServerIfStarted
+    case ServerThreadFailed(error) => handleFatalError(error)
+    case _                         => stash()
+  }
+
+  private def handleFatalError(error: Throwable): Unit = {
+    context.parent ! ServerThreadFailed(error)
+    context.stop(self)
+  }
+
+  /** In booting stage, the actor is retrying to connect to the server to verify
+    * that it has finished initialization.
+    *
+    * Once initialization is confirmed, it notifies the parent and proceeds to
+    * runningStage.
+    *
+    * Before initialization is confirmed the actor can also:
+    * - trigger the timeout or receive a graceful stop message which will ask
+    *   the child process to terminate gracefully,
+    * - get a notification that the child process has terminated spuriously
+    *   (possibly a crash).
+    */
+  private def bootingStage(
+    process: Process,
+    bootTimeout: Cancellable
+  ): Receive =
+    handleBootResponse(process, bootTimeout).orElse(runningStage(process))
+
+  case object AskServerIfStarted
+  case object TimedOut
+  private val retryDelay = 100.milliseconds
+
+  private def handleBootResponse(
+    process: Process,
+    bootTimeout: Cancellable
+  ): Receive = {
+    case AskServerIfStarted =>
+      val socket = Socket(descriptor.networkConfig.interface, rpcPort)
+      context.actorOf(
+        HeartbeatSession.initialProps(
+          socket,
+          retryDelay,
+          new AkkaBasedWebSocketConnectionFactory()(context.system),
+          context.system.scheduler
+        ),
+        s"initial-heartbeat-${UUID.randomUUID()}"
+      )
+    case TimedOut =>
+      self ! Stop
+    case HeartbeatReceived =>
+      context.parent ! LanguageServerProcess.ServerConfirmedFinishedBooting
+      bootTimeout.cancel()
+      context.become(runningStage(process))
+    case ServerUnresponsive =>
+      import context.dispatcher
```

Comment:
Please move this import higher on the class level. It is an unnecessary repetition.

Reply:
That was because I did not want to mix the 'compute' execution context for the forked process with this ordinary context. But once I move the logic to a separate class as you suggested it will indeed be better to move it up.